sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,226 @@
+   * allows for a static value to be provided.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class StaticValueProvider<T> implements ValueProvider<T>, Serializable {
+    @Nullable
+    private final T value;
+
+    StaticValueProvider(@Nullable T value) {
+      this.value = value;
+    }
+
+    /**
+     * Creates a {@link StaticValueProvider} that wraps the provided value.
+     */
+    public static <T> StaticValueProvider<T> of(T value) {
+      StaticValueProvider<T> factory = new StaticValueProvider<>(value);
+      return factory;
+    }
+
+    @Override
+    public T get() {
+      return value;
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return true;
+    }
+  }
+
+  /**
+   * {@link RuntimeValueProvider} is an implementation of {@link ValueProvider} that
+   * allows for a value to be provided at execution time rather than at graph
+   * construction time.
+   *
+   * <p>To enforce this contract, if there is no default, users must only call
+   * {@link #get()} at execution time (after a call to {@link Pipeline#run}),
+   * which will provide the value of OPTIONS.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class RuntimeValueProvider<T> implements ValueProvider<T>, Serializable {
+    private static ConcurrentHashMap<Long, PipelineOptions> optionsMap =
+      new ConcurrentHashMap<>();
+
+    private final Class<? extends PipelineOptions> klass;
+    private final String methodName;
+    private final T defaultValue;
+    private final Long optionsId;
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+                         Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = null;
+      this.optionsId = optionsId;
+    }
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value, or use the default if no value is available.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+      T defaultValue, Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = defaultValue;
+      this.optionsId = optionsId;
+    }
+
+    static void setRuntimeOptions(PipelineOptions runtimeOptions) {
+      optionsMap.put(runtimeOptions.getOptionsId(), runtimeOptions);
+    }
+
+    @Override
+    public T get() {
+      PipelineOptions options = optionsMap.get(optionsId);
+      if (options == null) {
+        if (defaultValue != null) {
+          return defaultValue;
+        }
+        throw new RuntimeException("Not called from a runtime context.");
```

Comment:
Thinking about this more, and I think having the default  be validated is the behavior we want --otherwise  it's a regression for most I/O things, right?